README.md
lv_font_conv - font convertor to compact bitmap format
======================================================

[![CI](https://github.com/lvgl/lv_font_conv/workflows/CI/badge.svg?branch=master)](https://github.com/lvgl/lv_font_conv/actions)
[![NPM version](https://img.shields.io/npm/v/lv_font_conv.svg?style=flat)](https://www.npmjs.org/package/lv_font_conv)

Converts TTF/WOFF fonts to __[compact format](https://github.com/lvgl/lv_font_conv/blob/master/doc/font_spec.md)__, suitable for small embedded systems. Main features are:

- Allows bitonal and anti-aliased glyphs (1-4 bits per pixel).
- Preserves kerning info.
- Compression.
- Users can select required glyphs only (subsetting).
- Multiple font sources can be merged.
- Simple CLI interface, easy to integrate into external build systems.


## Install the script

[node.js](https://nodejs.org/en/download/) v10+ required.

Global install of the last version, execute as "lv_font_conv"

```sh
# install release from npm registry
npm i lv_font_conv -g
# install from github's repo, master branch
npm i lvgl/lv_font_conv -g
```

**run via [npx](https://www.npmjs.com/package/npx) without install**

```sh
# run from npm registry
npx lv_font_conv -h
# run from github master
npx github:lvgl/lv_font_conv -h
```

Note, runing via `npx` may take some time until modules installed, be patient.


## CLI params

Common:

- `--bpp` - bits per pixel (antialiasing).
- `--size` - output font size (pixels).
- `-o`, `--output` - output path (file or directory, depends on format).
- `--format` - output format.
  - `--format dump` - dump glyph images and font info, useful for debug.
  - `--format bin` - dump font in binary form (as described in [spec](https://github.com/lvgl/lv_font_conv/blob/master/doc/font_spec.md)).
  - `--format lvgl` - dump font in [LittlevGL](https://github.com/lvgl/lvgl) format.
- `--force-fast-kern-format` - always use more fast kering storage format,
  at cost of some size. If size difference appears, it will be displayed.
- `--lcd` - generate bitmaps with 3x horizontal resolution, for subpixel
  smoothing.
- `--lcd-v` - generate bitmaps with 3x vertical resolution, for subpixel
  smoothing.
- `--use-color-info` - try to use glyph color info from font to create
  grayscale icons. Since gray tones are emulated via transparency, result
  will be good on contrast background only.
- `--lv-include` - only with `--format lvgl`, set alternate path for `lvgl.h`.

Per font:

- `--font` - path to font file (ttf/woff/woff2). May be used multiple time for
  merge.
- `-r`, `--range` - single glyph or range + optional mapping, belongs to
  previously declared `--font`. Can be used multiple times. Examples:
  - `-r 0x1F450` - single value, dec or hex format.
  - `-r 0x1F450-0x1F470` - range.
  - `-r '0x1F450=>0xF005'` - single glyph with mapping.
  - `-r '0x1F450-0x1F470=>0xF005'` - range with mapping.
  - `-r 0x1F450 -r 0x1F451-0x1F470` - 2 ranges.
  - `-r 0x1F450,0x1F451-0x1F470` - the same as above, but defined with single `-r`.
- `--symbols` - list of characters to copy (instead of numeric format in `-r`).
  - `--symbols 0123456789.,` - extract chars to display numbers.
- `--autohint-off` - do not force autohinting ("light" is on by default).
- `--autohint-strong` - use more strong autohinting (will break kerning).

Additional debug options:

- `--no-compress` - disable built-in RLE compression.
- `--no-prefilter` - disable bitmap lines filter (XOR), used to improve
  compression ratio.
- `--no-kerning` - drop kerning info to reduce size (not recommended).
- `--full-info` - don't shorten 'font_info.json' (include pixels data).


## Examples

Merge english from Roboto Regular and icons from Font Awesome, and show debug
info:

`env DEBUG=* lv_font_conv --font Roboto-Regular.ttf -r 0x20-0x7F --font FontAwesome.ttf -r 0xFE00=>0x81 --size 16 --format bin --bpp 3 --no-compress -o output.font`

Merge english & russian from Roboto Regular, and show debug info:

`env DEBUG=* lv_font_conv --font Roboto-Regular.ttf -r 0x20-0x7F -r 0x401,0x410-0x44F,0x451 --size 16 --format bin --bpp 3 --no-compress -o output.font`

Dump all Roboto glyphs to inspect icons and font details:

`lv_font_conv --font Roboto-Regular.ttf -r 0x20-0x7F --size 16 --format dump --bpp 3 -o ./dump`

**Note**. Option `--no-compress` exists temporary, to avoid confusion until LVGL
adds compression support.


## Technical notes

### Supported output formats

1. **bin** - universal binary format, as described in https://github.com/lvgl/lv_font_conv/tree/master/doc.
2. **lvgl** - format for LittlevGL, C file. Has minor limitations and a bit
   bigger size, because C does not allow to effectively define relative offsets
   in data blocks.
3. **dump** - create folder with each glyph in separate image, and other font
   data as `json`. Useful for debug.

### Merged font metrics

When multiple fonts merged into one, sources can have different metrics. Result
will follow principles below:

1. No scaling. Glyphs will have exactly the same size, as intended by font authors.
2. The same baseline.
3. `OS/2` metrics (`sTypoAscender`, `sTypoDescender`, `sTypoLineGap`) will be
   used from the first font in list.
4. `hhea`  metrics (`ascender`, `descender`), defined as max/min point of all
   font glyphs, are recalculated, according to new glyphs set.


## Development

Current package includes WebAssembly build of FreeType with some helper
functions. Everything is wrapped into Docker and requires zero knowledge about
additional tools install. See `package.json` for additional commands. You may
need those if decide to upgrade FreeType or update helpers.

This builds image with emscripten & freetype, usually should be done only once:

```
npm run build:dockerimage
```

This compiles helpers and creates WebAssembly files:

```
npm run build:freetype
```


CHANGELOG.md
1.5.2 / WIP
------------------

- Fixed lvgl version check for v8+, #64.


1.5.1 / 2021-04-06
------------------

- Fixed fail of CMAP generation for edge cases, #62.
- Dev deps bump.


1.5.0 / 2021-03-08
------------------

- More `const` in generated font (for v8+), #59.


1.4.1 / 2021-01-26
------------------

- Fix charcodes padding in comments, #54.


1.4.0 / 2021-01-03
------------------

- Added OTF fonts support.
- Added `--use-color-info` for limited multi-tone glyphs support.


1.3.1 / 2020-12-28
------------------

- Unify `lvgl.h` include.
- Updated repo refs (littlevgl => lvgl).
- Deps bump.
- Moved CI to github actions.


1.3.0 / 2020-10-25
------------------

- Drop `lodash` use.
- Deps bump.


1.2.1 / 2020-10-24
------------------

- Reduced npm package size (drop unneeded files before publish).


1.2.0 / 2020-10-24
------------------

- Bump FreeType to 2.10.4.
- Bundle dependencies to npm package.


1.1.3 / 2020-09-22
------------------

- lvgl: added `LV_FONT_FMT_TXT_LARGE` check or very large fonts.


1.1.2 / 2020-08-23
------------------

- Fix: skip `glyph.advanceWidth` for monospace fonts, #43.
- Spec fix: version size should be 4 bytes, #44.
- Spec fix: bbox x/y bits => unsigned, #45.
- Bump argparse.
- Cleanup help formatter.


1.1.1 / 2020-08-01
------------------

- `--version` should show number from `package.json`.


1.1.0 / 2020-07-27
------------------

- Added `post.underlinePosition` & `post.underlineThickness` info to font header.


1.0.0 / 2020-06-26
------------------

- Maintenance release.
- Set package version 1.x, to label package as stable.
- Deps bump.


0.4.3 / 2020-03-05
------------------

- Enabled `--bpp 8` mode.


0.4.2 / 2020-01-05
------------------

- Added `--lv_include` option to set alternate `lvgl.h` path.
- Added guards to hide `.subpx` property for lvgl 6.0 (supported from 6.1 only), #32.
- Dev deps bump


0.4.1 / 2019-12-09
------------------

- Allow memory growth for FreeType build, #29.
- Dev deps bump.
- Web build update.


0.4.0 / 2019-11-29
------------------

- Note, this release is for lvgl 6.1 and has potentially breaking changes
  (see below). If you have compatibility issues with lvgl 6.0 - use previous
  versions or update your code.
- Spec change: added subpixels info field to font header (header size increased).
- Updated `bin` & `lvgl` writers to match new spec.
- lvgl: fixed data type for kerning values (needs appropriate update
  in LittlevGL 6.1+).
- Fix errors display (disable emscripten error catcher).


0.3.1 / 2019-10-24
------------------

- Fixed "out of range" error for big `--size`.


0.3.0 / 2019-10-12
------------------

- Added beta options `--lcd` & `--lcd-v` for subpixel rendering (still need
  header info update).
- Added FreeType data properties to dump info.
- Fixed glyph width (missed fractional part after switch to FreeType).
- Fixed missed sigh for negative X/Y bitmap offsets.
- Deps bump.


0.2.0 / 2019-09-26
------------------

- Use FreeType renderer. Should solve all regressions, reported in 0.1.0.
- Enforced light autohinting (horizontal lines only).
- Use special hinter for monochrome output (improve quality).
- API changed to async.
- Fix: added missed `.bitmap_format` field to lvgl writer.
- Fix: changed struct fields init order to match declaration, #25.


0.1.0 / 2019-09-03
------------------

- First release.


lib/cli.js
// Parse input arguments and execute convertor

'use strict';


const argparse    = require('argparse');
const fs          = require('fs');
const mkdirp      = require('mkdirp');
const path        = require('path');
const convert     = require('./convert');


class ActionFontAdd extends argparse.Action {
  call(parser, namespace, value/*, option_string*/) {
    let items = (namespace[this.dest] || []).slice();
    items.push({ source_path: value, ranges: [] });
    namespace[this.dest] = items;
  }
}


// add range or symbols to font;
// need to merge them into one array here so overrides work correctly
class ActionFontRangeAdd extends argparse.Action {
  call(parser, namespace, value, option_string) {
    let fonts = namespace.font || [];

    if (fonts.length === 0) {
      parser.error(`argument ${option_string}: Only allowed after --font`);
    }

    let lastFont = fonts[fonts.length - 1];

    // { symbols: 'ABC' }, or { range: [ 65, 67, 65 ] }
    lastFont.ranges.push({ [this.dest]: value });
  }
}


// add hinting option to font;
class ActionFontStoreTrue extends argparse.Action {
  constructor(options) {
    options = options || {};
    options.const = true;
    options.default = options.default !== null ? options.default : false;
    options.nargs = 0;
    super(options);
  }

  call(parser, namespace, value, option_string) {
    let fonts = namespace.font || [];

    if (fonts.length === 0) {
      parser.error(`argument ${option_string}: Only allowed after --font`);
    }

    let lastFont = fonts[fonts.length - 1];

    lastFont[this.dest] = this.const;
  }
}


// Formatter with support of `\n` in Help texts.
class RawTextHelpFormatter2 extends argparse.RawDescriptionHelpFormatter {
  // executes parent _split_lines for each line of the help, then flattens the result
  _split_lines(text, width) {
    return [].concat(...text.split('\n').map(line => super._split_lines(line, width)));
  }
}


// parse decimal or hex code in unicode range
function unicode_point(str) {
  let m = /^(?:(?:0x([0-9a-f]+))|([0-9]+))$/i.exec(str.trim());

  if (!m) throw new TypeError(`${str} is not a number`);

  let [ , hex, dec ] = m;

  let value = hex ? parseInt(hex, 16) : parseInt(dec, 10);

  if (value > 0x10FFFF) throw new TypeError(`${str} is out of unicode range`);

  return value;
}


// parse range
function range(str) {
  let result = [];

  for (let s of str.split(',')) {
    let m = /^(.+?)(?:-(.+?))?(?:=>(.+?))?$/i.exec(s);

    let [ , start, end, mapped_start ] = m;

    if (!end) end = start;
    if (!mapped_start) mapped_start = start;

    start = unicode_point(start);
    end = unicode_point(end);

    if (start > end) throw new TypeError(`Invalid range: ${s}`);

    mapped_start = unicode_point(mapped_start);

    result.push(start, end, mapped_start);
  }

  return result;
}


// exclude negative numbers and non-numbers
function positive_int(str) {
  if (!/^\d+$/.test(str)) throw new TypeError(`${str} is not a valid number`);

  let n = parseInt(str, 10);

  if (n <= 0) throw new TypeError(`${str} is not a valid number`);

  return n;
}


module.exports.run = async function (argv, debug = false) {

  //
  // Configure CLI
  //

  let parser = new argparse.ArgumentParser({
    add_help: true,
    formatter_class: RawTextHelpFormatter2
  });

  if (debug) {
    parser.exit = function (status, message) {
      throw new Error(message);
    };
  }

  parser.add_argument('-v', '--version', {
    action: 'version',
    version: require('../package.json').version
  });

  parser.add_argument('--size', {
    metavar: 'PIXELS',
    type: positive_int,
    required: true,
    help: 'Output font size, pixels.'
  });

  parser.add_argument('-o', '--output', {
    metavar: '<path>',
    help: 'Output path.'
  });

  parser.add_argument('--bpp', {
    choices: [ 1, 2, 3, 4, 8 ],
    type: positive_int,
    required: true,
    help: 'Bits per pixel, for antialiasing.'
  });

  let lcd_group = parser.add_mutually_exclusive_group();

  lcd_group.add_argument('--lcd', {
    action: 'store_true',
    default: false,
    help: 'Enable subpixel rendering (horizontal pixel layout).'
  });

  lcd_group.add_argument('--lcd-v', {
    action: 'store_true',
    default: false,
    help: 'Enable subpixel rendering (vertical pixel layout).'
  });

  parser.add_argument('--format', {
    choices: convert.formats,
    required: true,
    help: 'Output format.'
  });

  parser.add_argument('--font', {
    metavar: '<path>',
    action: ActionFontAdd,
    required: true,
    help: 'Source font path. Can be used multiple times to merge glyphs from different fonts.'
  });

  parser.add_argument('-r', '--range', {
    type: range,
    action: ActionFontRangeAdd,
    help: `
Range of glyphs to copy. Can be used multiple times, belongs to previously declared "--font". Examples:
  -r 0x1F450
  -r 0x20-0x7F
  -r 32-127
  -r 32-127,0x1F450
  -r '0x1F450=>0xF005'
  -r '0x1F450-0x1F470=>0xF005'
`
  });

  parser.add_argument('--symbols', {
    action: ActionFontRangeAdd,
    help: `
List of characters to copy, belongs to previously declared "--font". Examples:
  --symbols ,.0123456789
  --symbols abcdefghigklmnopqrstuvwxyz
`
  });

  parser.add_argument('--autohint-off', {
    type: range,
    action: ActionFontStoreTrue,
    help: 'Disable autohinting for previously declared "--font"'
  });

  parser.add_argument('--autohint-strong', {
    type: range,
    action: ActionFontStoreTrue,
    help: 'Use more strong autohinting for previously declared "--font" (will break kerning)'
  });

  parser.add_argument('--force-fast-kern-format', {
    dest: 'fast_kerning',
    action: 'store_true',
    default: false,
    help: 'Always use kern classes instead of pairs (might be larger but faster).'
  });

  parser.add_argument('--no-compress', {
    dest: 'no_compress',
    action: 'store_true',
    default: false,
    help: 'Disable built-in RLE compression.'
  });

  parser.add_argument('--no-prefilter', {
    dest: 'no_prefilter',
    action: 'store_true',
    default: false,
    help: 'Disable bitmap lines filter (XOR), used to improve compression ratio.'
  });

  parser.add_argument('--no-kerning', {
    dest: 'no_kerning',
    action: 'store_true',
    default: false,
    help: 'Drop kerning info to reduce size (not recommended).'
  });

  parser.add_argument('--lv-include', {
    metavar: '<path>',
    help: 'Set alternate "lvgl.h" path (for --format lvgl).'
  });

  parser.add_argument('--full-info', {
    dest: 'full_info',
    action: 'store_true',
    default: false,
    help: 'Don\'t shorten "font_info.json" (include pixels data).'
  });

  //
  // Process CLI options
  //

  let args = parser.parse_args(argv.length ? argv : [ '-h' ]);

  for (let font of args.font) {
    if (font.ranges.length === 0) {
      parser.error(`You need to specify either "--range" or "--symbols" for font "${font.source_path}"`);
    }

    try {
      font.source_bin = fs.readFileSync(font.source_path);
    } catch (err) {
      parser.error(`Cannot read file "${font.source_path}": ${err.message}`);
    }
  }

  //
  // Convert
  //

  let files = await convert(args);

  //
  // Store files
  //

  for (let [ filename, data ] of Object.entries(files)) {
    let dir = path.dirname(filename);

    mkdirp.sync(dir);

    fs.writeFileSync(filename, data);
  }

};


// export for tests
module.exports._range = range;
